modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionDemander.java

```diff
@@ -1707,70 +1762,75 @@ public boolean compatibleWith(GridDhtPreloaderAssignments otherAssignments) {
         /**
          * Collect demander per cache groups. For print statistics.
          *
-         * @return List demanders.
+         * @param demanderPred Demander predicate.
+         * @return Set demanders.
          */
-        private Set<GridDhtPartitionDemander> demanders() {
+        private Set<GridDhtPartitionDemander> demanders(Predicate<? super GridDhtPartitionDemander> demanderPred) {
             return ctx.cacheContexts().stream()
                 .map(GridCacheContext::preloader)
                 .filter(GridDhtPreloader.class::isInstance)
                 .map(GridDhtPreloader.class::cast)
                 .map(GridDhtPreloader::demander)
+                .filter(demanderPred)
                 .collect(toSet());
         }
 
         /**
-         * Print rebalance statistics into log.
-         * Statistic will print if
-         * {@link RebalanceStatisticsUtils#printRebalanceStatistics()
-         * printRebalanceStatistics()} == true.
-         * To use correctly you need to call this method exactly once right after
-         * {@code RebalanceFuture} was completed (successfully or not).
-         * <p/>
-         * If {@link #rebalanceFut} was done successfully, prints statistics
-         * for cache group.
-         * <p/>
-         * If the whole rebalance is over, print statistics for all cache groups.
-         * The end of the rebalance is determined by the successful done all
-         * {@code RebalanceFuture}'s.
+         * Logging statistics of rebalance. Statistics will be printed for
+         * cache group and total if rebalance has ended for all cache groups
+         * successfully.
          *
-         * @throws IgniteCheckedException when get result {@code RebalanceFuture}
-         * @see RebalanceFuture RebalanceFuture
+         * @throws IgniteCheckedException If error occurs.
          */
         private void printRebalanceStatistics() throws IgniteCheckedException {
-            if (!RebalanceStatisticsUtils.printRebalanceStatistics())
-                return;
-
             assert isDone() : "RebalanceFuture should be done.";
```

Comment:
Can we assert that stat and totalRebStat are not null as well?
They are marked as NPE-prone by IDEA.

Reply:
Additionally check or replace?

Reply:
Below isDone() we can add additional assertions.